File: server_integration_tests/servers/java_server/src/main/java/dev/graphlink/javaserver/impl/MessageServiceImpl.java
package dev.graphlink.javaserver.impl;

import dev.graphlink.javaserver.generated.services.MessageReadSchemaMappingsService;
import dev.graphlink.javaserver.generated.services.MessageService;
import dev.graphlink.javaserver.generated.types.Message;
import org.springframework.stereotype.Service;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@Service
public class MessageServiceImpl implements MessageService, MessageReadSchemaMappingsService {

    @Override
    public List<Message> getMessageReadList() {
        return List.of(
            Message.builder().id("m1").content("hello").build(),
            Message.builder().id("m2").content("world").build());
    }

    @Override
    public Map<Message, Boolean> messageReadRead(List<Message> value) {
        Map<Message, Boolean> result = new LinkedHashMap<>();
        for (Message m : value) {
            result.put(m, m.getId().endsWith("1"));
        }
        return result;
    }
}

File: server_integration_tests/servers/kotlin_server_blocking/src/main/kotlin/dev/graphlink/kotlinserverblocking/impl/MessageServiceImpl.kt
package dev.graphlink.kotlinserverblocking.impl

import dev.graphlink.kotlinserverblocking.generated.services.MessageReadSchemaMappingsService
import dev.graphlink.kotlinserverblocking.generated.services.MessageService
import dev.graphlink.kotlinserverblocking.generated.types.Message
import org.springframework.stereotype.Service

@Service
class MessageServiceImpl : MessageService, MessageReadSchemaMappingsService {

    override fun getMessageReadList(): List<Message?>? = listOf(
        Message(id = "m1", content = "hello"),
        Message(id = "m2", content = "world"),
    )

    override fun messageReadRead(value: List<Message>): Map<Message, Boolean> {
        val result = LinkedHashMap<Message, Boolean>()
        for (m in value) {
            result[m] = m.id.endsWith("1")
        }
        return result
    }
}
